utilsforecast/feature_engineering.py

```diff
@@ -295,10 +295,61 @@ def time_features(
     return transformed, future
 
 # %% ../nbs/feature_engineering.ipynb 19
+def future_exog_to_historic(
+    df: DFType,
+    freq: Union[str, int],
+    features: List[str],
+    h: int = 0,
+    id_col: str = "unique_id",
+    time_col: str = "ds",
+) -> Tuple[DFType, DFType]:
+    """Turn future exogenous features into historic
+
+    Parameters
+    ----------
+    df : pandas or polars DataFrame
+        Dataframe with ids, times and values for the exogenous regressors.
+    freq : str or int
+        Frequency of the data. Must be a valid pandas or polars offset alias, or an integer.
+    features : list of str
+        Features to be converted into historic.
+    h : int (default=0)
+        Forecast horizon.
+    id_col : str (default='unique_id')
+        Column that identifies each serie.
+    time_col : str (default='ds')
+        Column that identifies each timestep, its values can be timestamps or integers.
+
+    Returns
+    -------
+    transformed_df : pandas or polars DataFrame
+        Original DataFrame with the computed features
+    future_df : pandas or polars DataFrame
+        DataFrame with future values
+    """
+    if h == 0:
+        return df, type(df)({})
+    last_times_by_id = ufp.group_by_agg(
+        df, id_col, {time_col: "max"}, maintain_order=True
+    )
+    last_times_by_id = ufp.sort(last_times_by_id, id_col)
+    future = ufp.make_future_dataframe(
+        uids=last_times_by_id[id_col],
+        last_times=last_times_by_id[time_col],
+        freq=freq,
+        h=h,
+        id_col=id_col,
+        time_col=time_col,
+    )
+    last_obs_by_id = ufp.join(last_times_by_id, df, on=[id_col, time_col])
+    future = ufp.join(future, last_obs_by_id[[id_col, *features]], on=id_col)
+    return df, future
```

Comment:
- How to use this with numpy arrays in the TimeGPT API because we don't want to do this in the SDK. And it's expensive to convert everything back to DataFrames.
- This is somewhat expensive as it requires a groupby and a sort, which isn't necessary if we directly operate on nd arrays.